src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/SystemNetworkInterface.cs

```diff
@@ -0,0 +1,261 @@
+                        while (ptr != IntPtr.Zero)
+                        {
+                            // Traverse the list, marshal in the native structures, and create new NetworkInterfaces.
+                            Interop.IpHlpApi.IpAdapterAddresses adapterAddresses = Marshal.PtrToStructure<Interop.IpHlpApi.IpAdapterAddresses>(ptr);
+                            interfaceList.Add(new SystemNetworkInterface(fixedInfo, adapterAddresses));
+
+                            ptr = adapterAddresses.next;
+                        }
+                    }
+                }
+                finally
+                {
+                    if (buffer != null)
+                    {
+                        buffer.Dispose();
+                    }
+
+                    buffer = null;
+                }
+            }
+
+            // If we don't have any interfaces detected, return empty.
+            if (result == Interop.IpHlpApi.ERROR_NO_DATA || result == Interop.IpHlpApi.ERROR_INVALID_PARAMETER)
+            {
+                return new SystemNetworkInterface[0];
+            }
+
+            // Otherwise we throw on an error.
+            if (result != Interop.IpHlpApi.ERROR_SUCCESS)
+            {
+                throw new NetworkInformationException((int)result);
+            }
+
+            return interfaceList.ToArray();
+        }
+
+        internal SystemNetworkInterface(Interop.IpHlpApi.FIXED_INFO fixedInfo, Interop.IpHlpApi.IpAdapterAddresses ipAdapterAddresses)
+        {
+            // Store the common API information.
+            _id = ipAdapterAddresses.AdapterName;
+            _name = ipAdapterAddresses.friendlyName;
+            _description = ipAdapterAddresses.description;
+            _index = ipAdapterAddresses.index;
+
+            _physicalAddress = ipAdapterAddresses.address;
+            _addressLength = ipAdapterAddresses.addressLength;
+
+            _type = ipAdapterAddresses.type;
+            _operStatus = ipAdapterAddresses.operStatus;
+            _speed = (long)ipAdapterAddresses.receiveLinkSpeed;
+
+            // API specific info.
+            _ipv6Index = ipAdapterAddresses.ipv6Index;
+
+            _adapterFlags = ipAdapterAddresses.flags;
+            _interfaceProperties = new SystemIPInterfaceProperties(fixedInfo, ipAdapterAddresses);
+        }
+
+        public override string Id { get { return _id; } }
+
+        public override string Name { get { return _name; } }
+
+        public override string Description { get { return _description; } }
+
+        public override PhysicalAddress GetPhysicalAddress()
+        {
+            byte[] newAddr = new byte[_addressLength];
+
+            // Array.Copy only supports int and long while addressLength is uint (see IpAdapterAddresses).
+            // Will throw OverflowException if addressLength > Int32.MaxValue.
+            Array.Copy(_physicalAddress, newAddr, checked((int)_addressLength));
```

Comment:
It's slightly more efficient to also pass in the lower bounds of 0, rather than forcing Array.Copy to query for them.